doc/book/app_server/luajit_memprof.rst

```diff
@@ -0,0 +1,524 @@
+
+    local function concat(a)
+      local nstr = a.."a"
+      return nstr
+    end
+
+    local function format(a)
+      local nstr = string.format("%sa", a)
+      return nstr
+    end
+
+    collectgarbage() -- Clean up.
+
+    local binfile = "/tmp/memprof_"..(arg[0]):match("([^/]*).lua")..".bin"
+
+    local st, err = misc.memprof.start(binfile)
+    assert(st, err)
+
+    -- Payload.
+    for i = 1, 10000 do
+      local f = format(i)
+      local c = concat(i)
+    end
+    collectgarbage() -- Clean up.
+
+    local st, err = misc.memprof.stop()
+    assert(st, err)
+
+    os.exit()
+
+When you run this code :ref:`under Tarantool <profiler_usage_generate>` and
+then :ref:`parse <profiler_usage_parse_command>` the binary memory profile,
+you will get the following profiling report:
+
+..  code-block:: console
+
+    ALLOCATIONS
+    @format_concat.lua:8, line 9: 19998     624322  0
+    INTERNAL: 1     65536   0
+
+    REALLOCATIONS
+
+    DEALLOCATIONS
+    INTERNAL: 19998 0       558816
+            Overrides:
+                    @format_concat.lua:8, line 9
+
+    @format_concat.lua:8, line 9: 2 0       98304
+            Overrides:
+                    @format_concat.lua:8, line 9
+
+The reasonable questions regarding the report can be:
+
+*   Why are there no allocations related to the ``concat()`` function?
+*   Why the amount of allocations is not a round number?
+*   Why are there approximately 20K allocations instead of 10K?
+
+First of all, LuaJIT doesn't create a new string if the string with the same
+payload exists. This is called the string interning. So, when the string is
+created via the ``format()`` function, there is no need to create the same
+string via the ``concat()`` function, and LuaJIT just use the previous one.
+
+That is also the reason why the amount of allocations is not the round numbber
+as can be expected from the cycle operator ``for i = 1, 10000...``:
+Tarantool creates some
+strings for internal needs and built-in modules, so some strings already exist.
+
+But why are there so many allocations? It's almost twice as big as the expected
+amount. This is because the ``string.format()`` built-in function creates
+another string necessary for the ``%s`` identifier, so there are two allocations
+for each iteration: for ``tostring(i)`` and for ``string.format("%sa", string_i_value)``.
+You can see the difference in behaviour by adding the
+``local _ = tostring(i)`` line between lines 21 and 22.
+
+To profile only the ``concat()`` function, comment the line 22, namely,
+``local f = format(i)`` and run the profiler.
+
+The profiler's output is the following:
+
+..  code-block:: console
+
+    ALLOCATIONS
+    @format_concat.lua:3, line 4: 10000     284411  0
+
+    REALLOCATIONS
+
+    DEALLOCATIONS
+    INTERNAL: 10000 0       218905
+            Overrides:
+                    @format_concat.lua:3, line 4
+
+    @format_concat.lua:3, line 4: 1 0       32768
+
```

Comment:
With the additional first line (see the comment above) the report looks like the following:
```
ALLOCATIONS
@format_concat.lua:4, line 5: 10000     284411  0

REALLOCATIONS

DEALLOCATIONS
INTERNAL: 10000 0       218905
        Overrides:
                @format_concat.lua:4, line 5
@format_concat.lua:4, line 5: 1 0       32768

```

Reply:
done